File: common/src/commander/cmd_factory.cpp
//
// Created by liuke on 15/11/6.
//

#include <commander/cmd_factory.h>
#include <util/UUID.h>
#include <iostream>
#include "log/log_wrap.h"

namespace plan9
{
    class cmd_factory::cmd_factory_impl
    {
    public:

        cmd_factory_impl() {

        }


        void register_cmd(std::string cmd, std::function<void(JSONObject)> function) {
            cmd_map_[cmd] = function;
        }

        void call(std::string cmd, JSONObject json, std::function<void(JSONObject)> callback) {
            JSONObject jsonWrap;
            if (!json.has("args") || !json.has("aux")) {
                jsonWrap["args"] = json;
            } else {
                jsonWrap = json;
            }
            std::string id = UUID::id();
            if (!jsonWrap["aux"].has("id")) {
                jsonWrap["aux"]["id"] = id;
            }
            jsonWrap["aux"]["to"] = cmd;
            if (callback != nullptr) {
                jsonWrap["aux"]["action"] = "callback";
                if (jsonWrap["aux"].has("from") && jsonWrap["aux"]["from"].is_array()) {
                    jsonWrap["aux"]["from"].append(id);
                    tmp_cmd_map_[id] = callback;
                } else {
                    jsonWrap["aux"]["from"].append(jsonWrap["aux"]["id"]);
                    tmp_cmd_map_[jsonWrap["aux"]["id"].get_string()] = callback;
                }


                if (cmd_map_.find(cmd) == cmd_map_.end()) {
                    //没有找到注册函数
                    jsonWrap["result.success"] = false;
                    jsonWrap["result.error"] = "CPP_FUNCTION_NOT_EXIST";
                    jsonWrap["result.reason"] = "the function : " + cmd + " is not exist";
                    callback(jsonWrap);
                    return;
                }
            } else {
                jsonWrap["aux"]["action"] = "direct";
            }
            execute_function(cmd, jsonWrap);
        }

        void call(JSONObject param, std::function<void(JSONObject)> callback) {
            if (param.has("aux") && param["aux"].has("to")) {
                std::string method = param["aux"]["to"].get_string();
                call(method, param, callback);
            }
        }

        void callback(JSONObject json, bool once) {
            using namespace std;
            if (json.has("aux")) {
                string action = json["aux"]["action"].get_string();
                if ("callback" == action) {
                    JSONObject from = json["aux.from"];
                    if (from.is_array() && from.size() > 0) {
                        if (!once || (json["aux"].has("once") && !(json["aux.once"].get_bool()))) {
                            //执行多次
                            bool find = false;
                            JSONObject fv = from.get(from.size() - 1, &find);
                            if (find) {
                                string f = fv.get_string();
                                if (tmp_cmd_map_.find(f) != tmp_cmd_map_.end()) {
                                    tmp_cmd_map_[f](json);
                                }
                            }
                        } else {
                            JSONObject fv = from.remove_last();
                            string f = fv.get_string();

                            if (tmp_cmd_map_.find(f) != tmp_cmd_map_.end()) {
                                tmp_cmd_map_[f](json);
                            }
                            tmp_cmd_map_.erase(f);
                        }
                    }
                }
            }
        }

        void callback(JSONObject json) {
            callback(json, true);
        }

        void execute_function(std::string cmd, JSONObject json) {
            if (cmd_map_.find(cmd) != cmd_map_.end()) {
                auto func = cmd_map_[cmd];
                func(json);
            } else {
                log_wrap::io().e("the c++ function : ", cmd , " is not exist");
            }
        }

        JSONObject wrap_callback(JSONObject json, bool result, JSONObject data) {
            json["result"]["success"] = result;
            json["result"]["data"] = data;
            return json;
        }

        bool is_register(std::string cmd) {
            return cmd_map_.find(cmd) != cmd_map_.end();
        }

    private:
        std::map<std::string, std::function<void(JSONObject)>> cmd_map_;
        std::map<std::string, std::function<void(JSONObject)>> tmp_cmd_map_;
    };


    cmd_factory::cmd_factory() : impl_(new cmd_factory_impl) {

    }

    cmd_factory cmd_factory::instance() {
        static cmd_factory cmd;
        return cmd;
    }

    void cmd_factory::register_cmd(std::string cmd, std::function<void(JSONObject)> function) {
        log_wrap::io().i("register function , method : ", cmd);
        impl_->register_cmd(cmd, function);
    }

    void cmd_factory::execute(std::string cmd, JSONObject param, std::function<void(JSONObject)> callback) {
        impl_->call(cmd, param, callback);
    }

    void cmd_factory::execute(JSONObject param, std::function<void(JSONObject)> callback) {
        impl_->call(param, callback);
    }


    void cmd_factory::callback(JSONObject json) {
        impl_->callback(json);
    }

    void cmd_factory::callback(JSONObject param, bool result, JSONObject data) {
        JSONObject ret = impl_->wrap_callback(param, result, data);
        callback(ret);
    }

    void cmd_factory::callback(JSONObject param, bool result) {
        callback(param, result, JSONObject());
    }

    void cmd_factory::execute(std::string cmd, JSONObject param) {
        impl_->call(cmd, param, nullptr);
    }


    void cmd_factory::execute(JSONObject param) {
        impl_->call(param, nullptr);
    }

    void cmd_factory::callback(JSONObject param, int error, std::string reason) {
        param["result.success"] = false;
        param["result.error"] = error;
        param["result.reason"] = reason;
        impl_->callback(param);
    }

    void cmd_factory::callback_multi(JSONObject param, bool result, int error, std::string reason, JSONObject data) {
        param["result.success"] = result;
        param["result.error"] = error;
        param["result.reason"] = reason;
        param["result.data"] = data;
        impl_->callback(param, false);
    }

    bool cmd_factory::is_register(std::string cmd) {
        return impl_->is_register(cmd);
    }

}

File: server_java/src/cn/gocoding/common/command/CmdFactory.java
package cn.gocoding.common.command;

import cn.gocoding.common.util.IDUtil;
import com.alibaba.fastjson.JSONArray;
import com.alibaba.fastjson.JSONObject;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.HashMap;
import java.util.Map;

/**
 * 以命令模式调用函数功能的类,记录命令,执行命令
 *
 * JSON对象的格式为:
 * {
 *    aux:  (辅助参数)
 *    {
 *         id:id                    (一个JSON对象调用的唯一标识)
 *         action:callback/direct   (callback,表示需要执行完成后最终执行要回到最初调用者的callback中,而direct则执行完成后不再回到之前的callback中)
 *         once:true/false          (true表示这个函数的callback会执行多次,false表示一次)
 *         from:[]                  (from是个数组,表示执行来源,最后根据这个来源再callback回去)
 *         to:cmd_name              (执行的命令名)
 *    }
 *    args:  (调用参数,这个结构中所有参数都是绑定函数所使用的)
 *    {
 *
 *    }
 *    result:   (调用函数最终返回的数据结构,为调用者使用)
 *    {
 *        success : true/false      (函数调用成功或者失败)
 *        error   :  error_code     这个错误码定义在lua的error_code文件中
 *        reason  :  失败原因        错误原因的文字说明,主要是UI使用
 *        data    : {}              (函数返回的结果数据封装在这个字段中)
 *    }
 *
 * }
 *
 * Created by liuke on 16/3/11.
 */
public class CmdFactory {

    /**
     * 注册一个函数名和执行的方法体
     * @param cmd 函数名
     * @param function 方法体
     */
    public static void registerCmd(String cmd, CmdFunction function) {
        cmdFunctionMap.put(cmd, function);
    }

    /**
     * 执行一个方法,并传入相应的参数.参数格式参考注释
     * @param cmd 函数名
     * @param param 参数
     * @param callback 回调
     */
    public static void execute(String cmd, JSONObject param, CmdCallback callback) {
        if (cmd == null || param == null) {
            throw new NullPointerException("执行命令名或者参数不能为空");
        } else {
            JSONObject aux = null;
            if (param.containsKey("aux")) {
                aux = param.getJSONObject("aux");
            } else {
                aux = new JSONObject();
                param.put("aux", aux);
            }

            aux.put("to", cmd);
            aux.put("id", IDUtil.createUniqueID());

            execute(param, callback);
        }

    }

    /**
     * 执行一个方法,参数中包含方法名和函数中使用的参数
     * @param param 参数
     */
    public static void execute(JSONObject param, CmdCallback callback) {
        if (param == null || !param.containsKey("aux") || !param.getJSONObject("aux").containsKey("to")) {
            throw new NullPointerException("执行的命令参数为空");
        } else {
            JSONObject aux = param.getJSONObject("aux");
            String to = aux.getString("to");
            String id = null;
            if (aux.containsKey("id")) {
                id = aux.getString("id");
            } else {
                id = IDUtil.createUniqueID();
                aux.put("id", id);
            }
            if (callback != null) {
                if (!aux.containsKey("action")) {
                    aux.put("action", "callback");
                }
                callbackMap.put(id, callback);
                if (aux.containsKey("from")) {
                    JSONArray from = aux.getJSONArray("from");
                    from.add(id);
                } else {
                    JSONArray from = new JSONArray();
                    from.add(id);
                    aux.put("from", from);
                }
            } else {
                if (!aux.containsKey("action")) {
                    aux.put("action", "direct");
                }
            }

            CmdFunction function = cmdFunctionMap.get(to);
            if (function != null) {
                function.run(param);
            } else {
                logger.error("can not execute the paramter : {} ,because of not find {} function, this cmd is not register", param, to);
            }
        }
    }

    /**
     * 回调调用者传入的callback接口函数,这个函数由命令实现函数体调用
     * @param result json格式数据
     */
    public static void callback(JSONObject result) {
        if (result == null) {
            throw new NullPointerException("回调参数不能为空");
        } else {
            if (result.containsKey("aux")) {
                JSONObject aux = result.getJSONObject("aux");
                JSONArray from = null;
                String action = null;
                if (aux.containsKey("from")) {
                    from = aux.getJSONArray("from");
                }
                if (aux.containsKey("action")) {
                    action = aux.getString("action");
                }

                if (from != null && action != null) {
                    if ("callback".equals(action) && !from.isEmpty()) {
                        aux.put("id", from.getString(0));
                        String lastID = from.getString(from.size() - 1);
                        CmdCallback callback = callbackMap.get(lastID);
                        if (callback != null) {
                            from.remove(from.size() - 1);
                            callback.callback(result);
                        }
                    }
                } else {
                    logger.error("parameter error, the from or action is null, result json : {}", result);
                }
            }
        }
    }

    /**
     * 返回结果成功的json数据
     * @param param 原始参数
     * @param data 成功后的数据实体,需要callback回调用者
     */
    public static void callback(JSONObject param, JSONObject data) {
        if (param == null) {
            throw new NullPointerException("回调的结果数据不能为空");
        } else {
            JSONObject result = null;
            if (param.containsKey("result")) {
                result = param.getJSONObject("result");
            } else {
                result = new JSONObject();
                param.put("result", result);
            }

            result.put("success", true);
            if (data == null) {
                param.put("data", new JSONObject());
            } else {
                param.put("data", data);
            }
            callback(param);
        }
    }

    /**
     * 返回失败结果的json数据
     * @param param 原始参数
     * @param errorCode 失败码
     * @param reason 失败原因描述
     */
    public static void callback(JSONObject param, int errorCode, String reason) {
        if (param == null) {
            throw new NullPointerException("回调的结果数据不能为空");
        } else {
            JSONObject result = null;
            if (param.containsKey("result")) {
                result = param.getJSONObject("result");
            } else {
                result = new JSONObject();
                param.put("result", result);
            }

            result.put("success", false);
            result.put("error", errorCode);
            result.put("reason", reason);

            callback(param);
        }
    }


    private static CmdFactory cmdFactory = new CmdFactory();
    private static Map<String, CmdFunction> cmdFunctionMap = new HashMap<>();
    private static Map<String, CmdCallback> callbackMap = new HashMap<>();

    private static final Logger logger = LogManager.getLogger(CmdFactory.class);
}
